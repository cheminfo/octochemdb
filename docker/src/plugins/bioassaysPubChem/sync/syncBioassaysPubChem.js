import { join } from 'path';

import { fileCollectionFromPath } from 'filelist-utils';
import md5 from 'md5';

import getLastDocumentImported from '../../../sync/http/utils/getLastDocumentImported.js';
import syncFolder from '../../../sync/http/utils/syncFolder.js';
import debugLibrary from '../../../utils/Debug.js';

import { decompressAll } from './utils/decompressAll.js';
import { main } from './utils/main.js';

const debug = debugLibrary('syncBioassaysPubChem');
export async function sync(connection) {
  let options = {
    collectionSource: process.env.BIOASSAYSPUBMECHEM_SOURCE,
    destinationLocal: `${process.env.ORIGINAL_DATA_PATH}/bioassaysPubChem/full`,
    collectionName: 'bioassaysPubChem',
  };
  try {
    const progress = await connection.getProgress(options.collectionName);
    let sources = [];
    if (process.env.NODE_ENV === 'test') {
      sources = [`${process.env.BIOASSAYSPUBMECHEM_SOURCE_TEST}`];
    } else {
      const { allFiles } = await syncFolder(
        options.collectionSource,
        options.destinationLocal,
        {
          fileFilter: (file) => file && file.name.endsWith('.zip'),
        },
      );
      if (allFiles.length > 0) {
        allFiles.forEach((file) => {
          sources.push(file?.path);
        });
      }
    }
    let isTimeToUpdate = false;
    if (
      progress.dateEnd !== 0 &&
      Date.now() - progress.dateEnd >
        Number(process.env.BIOASSAYPUBCHEM_UPDATE_INTERVAL) &&
      md5(JSON.stringify(sources)) !== progress.sources
    ) {
      progress.dateStart = Date.now();
      await connection.setProgress(progress);
      isTimeToUpdate = true;
    }

    const lastDocumentImported = await getLastDocumentImported(
      connection,
      progress,
      options.collectionName,
    );

    if (
      lastDocumentImported === null ||
      ((md5(JSON.stringify(sources)) !== progress.sources ||
        progress.state !== 'updated') &&
        isTimeToUpdate)
    ) {
      let sourceFiles = options.destinationLocal;
      if (process.env.NODE_ENV === 'test') {
        sourceFiles = process.env.BIOASSAYSPUBMECHEM_SOURCE_TEST || '';
      }
      let fileList = await fileCollectionFromPath(sourceFiles, {
        unzip: { zipExtensions: [] },
        ungzip: { gzipExtensions: [] },
      });
      let jsonFilesPaths = [];
      for (let file of fileList.files) {
        const newJsonFilesPaths = await decompressAll(
          join(sourceFiles, file.name),
        );
        jsonFilesPaths = [...newJsonFilesPaths, ...jsonFilesPaths];
      }
      // Generate Logs for the sync
      const logs = await connection.getImportationLog({
        collectionName: options.collectionName,
        sources,
        startSequenceID: progress.seq,
      });
      // set progress to updating, if fail importation, a new try will be done 24h later
      progress.state = 'updating';
      await connection.setProgress(progress);

      /// Start import
      await main(jsonFilesPaths);
      ///
      // Temporary collection to store the new data
      const temporaryCollection = await connection.getCollection(
        `${options.collectionName}_tmp`,
      );
      // Once it is finished, the temporary collection replace the old collection
      await temporaryCollection.rename(options.collectionName, {
        dropTarget: true,
      });
      // update logs with the new progress
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';
      await connection.updateImportationLog(logs);
      // update progress with the new progress
      progress.sources = md5(JSON.stringify(sources));
      progress.dateEnd = Date.now();
      progress.state = 'updated';
      await connection.setProgress(progress);
      // Indexing of properties in collection
      const collection = await connection.getCollection(options.collectionName);
      await collection.createIndex({ 'data.comment': 1 });
      await collection.createIndex({ 'data.description': 1 });
      await collection.createIndex({ 'data.name': 1 });
      await collection.createIndex({ 'data.results': 1 });
      await collection.createIndex({ 'data.sids': 1 });
      await collection.createIndex({ 'data.associatedCIDs': 1 });
    } else {
      debug.info(`collection already updated`);
    }
  } catch (e) {
    if (connection) {
      await debug.fatal(e.stack, {
        collection: options.collectionName,
        connection,
        stack: e.stack,
      });
    }
  }
}
