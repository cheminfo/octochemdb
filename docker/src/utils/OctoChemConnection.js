import delay from 'delay';
import dotenv from 'dotenv';
import md5 from 'md5';
import { MongoClient } from 'mongodb';

import debugLibrary from './Debug.js';

dotenv.config();
const debug = debugLibrary('OctoChemConnection');

export function OctoChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

OctoChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

OctoChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

OctoChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

OctoChemConnection.prototype.getImportationLog =
  async function getImportationLog(options) {
    const { collectionName, sources, startSequenceID } = options;
    const logsCollection = await this.getImportationLogsCollection();
    const sourcesHash = md5(JSON.stringify(sources));
    const _id = sourcesHash;
    let logs = await logsCollection.find({ _id }).next();
    if (logs === null) {
      logs = {
        _id,
        collectionName,
        sources,
        sourcesHash,
        dateStart: Date.now(),
        dateEnd: Date.now(),
        startSequenceID,
        endSequenceID: 0,
        status: 'updating',
      };
      await logsCollection.insertOne(logs);
    }
    return logs;
  };
OctoChemConnection.prototype.updateImportationLog =
  async function updateImportationLog(logs) {
    const collection = await this.getImportationLogsCollection();
    logs.dateEnd = Date.now();
    await collection.replaceOne({ _id: logs._id }, logs);
  };

OctoChemConnection.prototype.getImportationLogsCollection =
  async function getCollection() {
    return (await this.getDatabase()).collection('importationLogs');
  };

OctoChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

OctoChemConnection.prototype.setProgress = async function setProgress(
  progress,
) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

OctoChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'updating',
      seq: 0,
      dateStart: Date.now(),
      dateEnd: 0,
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

OctoChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

OctoChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
