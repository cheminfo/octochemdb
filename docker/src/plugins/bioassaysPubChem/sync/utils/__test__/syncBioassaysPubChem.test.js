import { test, expect } from 'vitest';

import { OctoChemConnection } from '../../../../../utils/OctoChemConnection.js';
import { sync } from '../../syncBioassaysPubChem';

test('syncBioassaysPubChem', async () => {
  const connection = new OctoChemConnection();

  await sync(connection);
  const collection = await connection.getCollection('bioassaysPubChem');
  const result = await collection.findOne({});
  console.log(result);
  // expect(result).toMatchSnapshot();
  await connection.close();
});
