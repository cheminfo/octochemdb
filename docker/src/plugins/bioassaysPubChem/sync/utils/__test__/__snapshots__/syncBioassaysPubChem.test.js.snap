// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`syncBioassaysPubChem 1`] = `
{
  "_id": 22001,
  "data": {
    "associatedCIDs": [
      44287812,
    ],
    "comment": [
      "Journal: Bioorg. Med. Chem. Lett.",
      "Year: 2002",
      "Volume: 12",
      "Issue: 12",
      "First Page: 1603",
      "Last Page: 1606",
      "DOI: 10.1016/s0960-894x(02)00226-3",
    ],
    "description": [
      "Title: Exploration of peptidyl hydrazones as water-soluble calpain inhibitors.",
      "Abstract: A series of peptidyl hydrazones was synthesized, and their inhibitory activity against mu-calpain and water-solubility were measured. Among these compounds, N,N-dimethyl glycyl hydrazone 6, which inhibited mu-calpain with IC(50) of 0.37 microM, possessed the appropriate water-solubility. Furthermore, hydrazone 6 was found to possess the excellent in vitro metabolic stability.",
    ],
    "name": "Water solubility was determined at pH 6; ND=Not determined",
    "results": [
      {
        "description": [
          {
            "description": [
              "Standardized activity type (e.g. IC50 rather than Ic-50/Ic50/ic50/ic-50)",
            ],
            "name": "Standard Type",
            "unit": "unspecified",
            "value": "Solubility",
          },
          {
            "description": [
              "Selected units for 'Standard Type': e.g. concentrations are in nM",
            ],
            "name": "Standard Units",
            "unit": "unspecified",
            "value": "ug.mL-1",
          },
          {
            "description": [
              "Additional comments",
            ],
            "name": "Activity Comment",
            "unit": "unspecified",
            "value": "Not Determined",
          },
        ],
        "outcome": "inconclusive",
        "sid": 103219760,
      },
    ],
    "sids": [
      103219760,
    ],
  },
}
`;
